nce_cost/README_en.md

```diff
@@ -0,0 +1,161 @@
+
+4. **FULL CONNECTION Hidden Layer**: Input the long vector obtained from the previous layer into a hidden layer of neural network and output the feature vector. A fully connected hidden layer can enhance the learning ability of the network.
+
+5. **NCE layer**: `The paddle.layer.nce` provided by PaddlePaddle can be used directly as a loss function during training.
+
+## Training
+Run the command ``` python train.py ``` in the command line window to start the training task.
+
+- The first time the program runs, it will detect if the ptb data set is included in the user's cache folder. If it is not included, it will be downloaded automatically.
+- Every 10 batch prints the value of model training on the training set during the run
+- After each pass is completed, the loss on the test data set is calculated and the latest snapshot of the model is also saved.
+
+The NCE call code in the model file `network_conf.py` is as follows:
+
+```python
+return paddle.layer.nce(
+            input=hidden_layer,
+            label=next_word,
+            num_classes=dict_size,
+            param_attr=paddle.attr.Param(name="nce_w"),
+            bias_attr=paddle.attr.Param(name="nce_b"),
+            num_neg_samples=25,
+            neg_distribution=None)
+```
+
+Some important parameters of the NCE layer are explained as follows:
+
+| Parameter name                    | Parameter function                                 | Introduction                                                                                                                                               |
+| :----------------------- | :--------------------------------------- | :--------------------------------------------------------------------------------------------------------------------------------------------------- |
+| param\_attr / bias\_attr | use to set parameter name  |  facilitates the loading of parameters during the forecast phase, as described in the prediction section                                                                                                   |
+| num\_neg\_samples        |  Number of Negative Samples                  |  You can control the proportion of positive and negative samples. The range of this value is [1, dictionary size -1]. The more negative samples, the slower the training of the whole model. The accuracy of the model will also be higher                                  |
+| neg\_distribution        |  Generates the distribution of negative sample tags. The default is a uniform distribution| You can control the sampling weights for each category when negative samples are sampled. For example, if you want the positive example to be “sunny,” and the negative example “flood” is highlighted more during training, you can increase the weight of the “flood” sample |
+
```

Comment:
# Accelerating Language Model Training with Noise Contrast Estimation

## Why We Need Noise Comparison

Language model is the basis of many natural language processing tasks, and it is also an effective method for obtaining word vector representations. The Neural Probabilistic Language Model (NPLM) characterizes the probability of a sequence of words $\omega_1,...,\omega_T$ belonging to a fixed language $P(\omega_1^T)$ :
$$P(\omega_1^T)= \prod_{t=1}^{T}P(\omega_t|\omega_1^{t-1})$$

In order to reduce the difficulty of modeling and solving, we often introduce certain conditional independent assumptions: the probability of the word $w_t$ is only affected by the previous $n-1$ words, so there is formula ($1$):

$$ P(\omega_1^T) \approx \prod P(\omega_t|\omega_{t-n-1}^{t-1}) \tag{1}$$

From the formula ($1$) we can see that the entire sequence $\omega_1,...,\omega_T$ can be calculated by the modeling conditional probabilities $P(\omega_t|w_{t-n-1},...,\omega_{t-1})$. So, we can summarize the task of language model solving as follows:

** Giver the vector representation or the context, $h$, of a word sequence, let the model predict the probability of the next target word $\omega$. **

In the [$n$-gram language model] (https://github.com/PaddlePaddle/book/tree/develop/04.word2vec), the context takes a fixed $n-1$ words. [RNN language model] (https://github.com/PaddlePaddle/models/tree/develop/generate_sequence_by_rnn_lm) can handle any length of context.

Given context $h$, NPLM learns a scoring function $s_\theta(\omega, h)$, which characterizes the similarities between the vector of $h$ and \omega'$, the vector of the possible words coming next. The similarity between \omega'$ is then normalized by dividing the value of the scoring function $s$ in the whole word table space (divided by the normalization factor $Z$) to get the probability distribution of the target word $\omega$. $\theta$ is a learning parameter, this process is expressed by formula ($2$), the calculation of `Softmax`.

$$P_\theta^h(\omega) = \frac{\text{exp}{s_\theta(\omega, h)}}{Z},Z=\sum_{\omega'} \exp{s_\ Theta(\omega', h)}\tag{2}$$

MLE (Maximum Likelihood Estimation) is the most commonly used learning criterion for solving probability ($2$). However, the normalization factor $Z$ must be calculated whether the estimated probability $P_\theta^h(\omega)$ or the gradient of the likelihood is calculated. The calculation of $Z$ increases linearly with the size of the dictionary. When training large-scale language models, such as a dictionary with a size of a million or more, the training time will become very large. Therefore, we need other possible learning guidelines with solution process easier and more solvable. **

Another article of models intruduces the use of [Hsigmoid Accelerated Word Vector Training] (https://github.com/PaddlePaddle/models/tree/develop/hsigmoid). Here, we introduce sample-based training model, Noise-contrastive estimation (NCE)\[[1](#ref)\], to speed up language training.

## What Noise Comparison Estimation Is

The noise comparison estimation is a probability density estimation criterion based on the sampling idea and is used for estimation/fitting: the probability function is composed of a non-normalized score function and a normalization factor[[1](#References)\]. The noise comparison estimation avoids the calculation of the normalization factor $Z$ in the full dictionary space by constructing an auxiliary problem such as the following, which reduces the computational cost:

Given context $h$ and any known noise distribution $P_n$, learn a two-class classifier to fit: target $\mega$ from true distribution $P_\theta$ ($D = 1$) or the probability of noise distribution, $P_n$ ($D = 0$). Assuming that the number of negative samples from the noise distribution is $k$ times the target sample, then there are:

$$P(D=1|h,\omega) = \frac{P_\theta(h, \omega)}{P_\theta (h, \omega) + kP_n} \tag{3}$$

We directly use the `Sigmoid` function to characterize a binary classification probability ($3$):

$$P(D=1|h,\omega) = \sigma (\Delta s_\theta(w,h)) \tag{4}$$

With the above problem set, maximum likelihood estimation can be performed based on two classifications: increasing the probability of positive samples and reducing the probability of negative samples [[2,3] (#references)], that is, minimizing the following loss function:

$$
J^h(\theta )=E_{ P_d^h }\left[ \log {P^h(D=1|w,\theta )} \right] +kE_{ P_n }\left[ \log P^h (D=0|w,\theta )\right]$$
$$
 \\\\\qquad =E_{ P_d^h }\left[ \log { \sigma (\Delta s_\theta(w,h)) } \right] +kE_{ P_n }\left[ \log (1- \sigma (\Delta s_\theta(w,h))) \right] \tag{5}$$

The formula ($5$) is the NCE loss function defined based on noise comparison estimation. At this point, we have two remaining questions:
1. What is $s_\theta(w,h)$ in the formula ($5$)?
    - In a neural network implementation, $s_\theta(h,\omega)$ is an unnormalized score.
    - The learnable parameter $W$ of the NCE cost layer is a matrix of $|V| \times d$ dimensions; $|V|$ is the dictionary size, and $d$ is the dimension of the context vector $h$;
    - During training, the actual category of the next word, $t$, is a positive class. Sampling $k$ negative samples from the specified noise distribution are classified as follows: $\{n_1, ..., n_k\}$;
    - Extract the $\{t, n_1, ..., n_k\}$th line in $W$ ($k + 1$ lines in total) and calculate the points $s_\theta(w,h)$ with $h$ respectively. The final loss is calculated by ($5$).
2. How to choose the noise distribution?
    - In practice, the noise distribution can be arbitrarily chosen (the noise distribution implies a certain priori).
    - The most common choices are: use the unbiased uniform distribution, `unigram`, based on a full dictionary (word frequency statistics).
    - If the user does not specify a noise distribution in the PaddlePaddle, the default is to use a uniform distribution.

When using NCE to accurately train, the calculation cost of the last layer is only linearly related to the number of negative samples. When the number of negative samples gradually increases, the NCE estimation criterion converges to the maximum likelihood estimation. Therefore, when training using the NCE criterion, the quality of the normalized probability distribution can be controlled by changing the number of negative samples.

## Experimental Data
This example uses a Penn Treebank (PTB) data set ([Tomas Mikolov pre-processed version] (http://www.fit.vutbr.cz/~imikolov/rnnlm/simple-examples.tgz)) to train a 5-gram language model. PaddlePaddle provides a [paddle.dataset.imikolov](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/v2/dataset/imikolov.py) interface to facilitate the use of PTB data. When no downloaded data is found, the script will automatically downloads and verifies the integrity of the file. The corpus language is in English, with a total of 42068 sentence training data and 3761 sentence test data.

## Network Structure
The detailed network structure of the 5-gram neural probabilistic language model is shown in Figure 1:

<p align="center">
<img src="images/network_conf.png" width = "70%" align="center"/><br/>
Figure 1. 5-gram network configuration structure
</p>

The model is mainly divided into the following parts:

1. **Input Layer**: the input sample consists of the original words, and each word is first converted to the ID in the dictionary.

2. **Word vector layer **: IDs are transformed to continously expressed word vector representation, which can better show the semantic relationship between words. After the training is completed, the semantic similarity between words can be expressed using the distance between the word vectors. The more similar the semantics, the closer the distance.

3. **Word vector splice layer**: concatenates the word vectors and connects the word vectors end to end to form a long vector. This can facilitate the processing of the fully connected layer.

4. **FULL CONNECTION Hidden Layer**: inputs the long vector obtained from the previous layer into a hidden layer of neural network and outputs the feature vector. A fully connected hidden layer can enhance the learning ability of the network.

5. **NCE layer**: The paddle.layer.nce` provided by PaddlePaddle can be used directly as a loss function during training.


## Training
Run ``` python train.py ``` in the command line window to start the training task.

- The first time the program runs, it will detect if the PTB data set is included in the user's cache folder. If it is not included, it will be automatically downloaded.
- Every 10 batch prints the value of model training on the training set during the run
- After each pass is completed, the loss on the test data set is calculated and the latest snapshot of the model is also saved.

The NCE code in the model file, `network_conf.py`, is as follows:

```python
Return paddle.layer.nce(
            Input=hidden_layer,
            Label=next_word,
            Num_classes=dict_size,
            Param_attr=paddle.attr.Param(name="nce_w"),
            Bias_attr=paddle.attr.Param(name="nce_b"),
            Num_neg_samples=25,
            Neg_distribution=None)
```

Some important parameters of the NCE layer are explained as follows:

Parameter Name | Parameter Function | Introduction |